### PR TITLE
Semantic checks for 5 data constraints: C874, C875, C878, C880, C881

### PR DESCRIPTION
**Pull request by anchu-rajendran**
This commit  has implementation of semantic checks for 5 constraints on data object. 3 more constraint-checks need to be implemented.

### CODE REVIEW

**Comment by anchu-rajendran** [comment]
Hi, 
I would like to get clarifications on two constraints
1. C878 (R841) The array-element shall be a variable.
To throw the error,  the only test case I could think of is the case where array element is a constant. So my code checks if data object is a constant and if yes, it throws the error. Please correct me if it has to be implemented in a different way.
2. C879 (R841) The scalar-structure-component shall be a variable.
According to my understanding, the code for C878 should apply for C879 as well (as it checks whether the `DataRef` is a constant or not). I tried writing a testcase where a scalar component of a struture is a constant with the snippet below:
`type newNum`
` integer a`
`end type`
`type(newNum), parameter :: num`
In this case, I see that num%a is not identified as a constant expression by F18. Shouldn't it be identified? If so, it requires modification in the implementation of `IsConstantExpr`
Please correct me if I am interpreting things wrong.
Thank you ,

**Comment by klausler** [comment]
> Hi,
> @klausler , @tskeith, I would like to get clarifications on two constraints
> 
> 1. C878 (R841) The array-element shall be a variable.
>    To throw the error,  the only test case I could think of is the case where array element is a constant. So my code checks if data object is a constant and if yes, it throws the error. Please correct me if it has to be implemented in a different way.

There's constraints (C924) on `array-element` as well - it needs to be a scalar (rank 0) `data-ref`.  For a variable, see C901, but yes, it just means that it cannot be an element of a named constant.  A `variable` can also be a reference to a function returning a data pointer, but I suspect that case is an error in the standard -- it couldn't be called at compilation time.

> 2. C879 (R841) The scalar-structure-component shall be a variable.
>    According to my understanding, the code for C878 should apply for C879 as well (as it checks whether the `DataRef` is a constant or not). I tried writing a testcase where a scalar component of a struture is a constant with the snippet below:
>    `type newNum`
>    ` integer a`
>    `end type`
>    `type(newNum), parameter :: num`
>    In this case, I see that num%a is not identified as a constant expression by F18. Shouldn't it be identified? If so, it requires modification in the implementation of `IsConstantExpr`

Every component in the derived type has to be initialized in the either the definition of the derived type or in the structure constructor used to initialized the named constant.  This code snippet is in error (C807); if we check C807 already, there's a bug.  `IsConstantExpr` is probably failing because there's no constant instance of the derived type to work with here.

>    Please correct me if I am interpreting things wrong.
>    Thank you ,

**Comment by anchu-rajendran** [comment]
> > Hi,
> > @klausler , @tskeith, I would like to get clarifications on two constraints
> > 
> > 1. C878 (R841) The array-element shall be a variable.
> >    To throw the error,  the only test case I could think of is the case where array element is a constant. So my code checks if data object is a constant and if yes, it throws the error. Please correct me if it has to be implemented in a different way.
> 
> There's constraints (C924) on `array-element` as well - it needs to be a scalar (rank 0) `data-ref`. For a variable, see C901, but yes, it just means that it cannot be an element of a named constant. A `variable` can also be a reference to a function returning a data pointer, but I suspect that case is an error in the standard -- it couldn't be called at compilation time.
> 
> > 1. C879 (R841) The scalar-structure-component shall be a variable.
> >    According to my understanding, the code for C878 should apply for C879 as well (as it checks whether the `DataRef` is a constant or not). I tried writing a testcase where a scalar component of a struture is a constant with the snippet below:
> >    `type newNum`
> >    ` integer a`
> >    `end type`
> >    `type(newNum), parameter :: num`
> >    In this case, I see that num%a is not identified as a constant expression by F18. Shouldn't it be identified? If so, it requires modification in the implementation of `IsConstantExpr`
> 
> Every component in the derived type has to be initialized in the either the definition of the derived type or in the structure constructor used to initialized the named constant. This code snippet is in error (C807); if we check C807 already, there's a bug. `IsConstantExpr` is probably failing because there's no constant instance of the derived type to work with here.
> 
> > Please correct me if I am interpreting things wrong.
> > Thank you ,

Hi,
Thank you for the response.
Sorry, I forgot to initialize it when I simplified the example for mentioning here.  The example I used for testing C879 is the following: 
```
type specialNumbers
  integer num
end type
type(specialNumbers), parameter ::newNumsArray(2) = (/ SpecialNumbers(1),  SpecialNumbers(2) /)
DATA (newNumsArray(i)%num, i = 1, 2) / 2 * specialNumbers(1) /
```
Here, I cheked whether `newNumsArray(i)%num` is a constant using `IsConstantExpr` and it was returning false.

**Comment by klausler** [comment]
> Here, I cheked whether `newNumsArray(i)%num` is a constant using `IsConstantExpr` and it was returning false.

That's probably because `IsConstantExpr` doesn't think that implied DO variables are constants; it probably works with a component reference to a scalar named constant, and it's a bug if it doesn't.

**Comment by anchu-rajendran** [comment]
> > Here, I cheked whether `newNumsArray(i)%num` is a constant using `IsConstantExpr` and it was returning false.
> 
> That's probably because `IsConstantExpr` doesn't think that implied DO variables are constants; it probably works with a component reference to a scalar named constant, and it's a bug if it doesn't.
Implied do variables are treated as constants by IsConstantExpr. Consider the following code snippet (which is semantically wrong again)  
```
type(specialNumbers), parameter ::newNum = SpecialNumbers(1)
DATA (newNum%num, i = 1, 2) / 2 * 1/
```
Here `IsConstantExpr` on `newNum%num`  also returns false.

**Comment by klausler** [comment]
> > > Here, I cheked whether `newNumsArray(i)%num` is a constant using `IsConstantExpr` and it was returning false.
> > 
> > 
> > That's probably because `IsConstantExpr` doesn't think that implied DO variables are constants; it probably works with a component reference to a scalar named constant, and it's a bug if it doesn't.
> > Implied do variables are treated as constants by IsConstantExpr. Consider the following code snippet (which is semantically wrong again)
> 
> ```
> type(specialNumbers), parameter ::newNum = SpecialNumbers(1)
> DATA (newNum%num, i = 1, 2) / 2 * 1/
> ```
> 
> Here `IsConstantExpr` on `newNum%num` also returns false.

Ok, thanks for checking.  That's a bug.  Want me to fix it for you, or would you like to do it?

**Comment by anchu-rajendran** [comment]
> > > > Here, I cheked whether `newNumsArray(i)%num` is a constant using `IsConstantExpr` and it was returning false.
> > > 
> > > 
> > > That's probably because `IsConstantExpr` doesn't think that implied DO variables are constants; it probably works with a component reference to a scalar named constant, and it's a bug if it doesn't.
> > > Implied do variables are treated as constants by IsConstantExpr. Consider the following code snippet (which is semantically wrong again)
> > 
> > 
> > ```
> > type(specialNumbers), parameter ::newNum = SpecialNumbers(1)
> > DATA (newNum%num, i = 1, 2) / 2 * 1/
> > ```
> > 
> > 
> > Here `IsConstantExpr` on `newNum%num` also returns false.
> 
> Ok, thanks for checking. That's a bug. Want me to fix it for you, or would you like to do it?

Thank you for the clarification. I will try to fix it and will get back if I have any doubts.

File: test/Semantics/data03.f90 [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      character :: name(12)
+      integer :: ind = 2
+      !C874
+      !ERROR: Data object must not be a coindexed variable
+      DATA a[1] / 1 /
+      !C874
+      !ERROR: Data object must not be a coindexed variable
+      DATA(a[i], i = 1, 5) / 5 * 1 /
+      !C875
+      !ERROR: Data object variable must be a designator
```
Comment (tskeith) [inline test/Semantics/data03.f90]: I suggest "Data object variable may not be a function reference". Users won't necessarily know what a designator is.

File: lib/Semantics/check-data.cpp [hunk 2]
```diff
@@ -10,6 +10,64 @@
 
 namespace Fortran::semantics {
 
+void DataChecker::CheckIfConstantExpr(const parser::Expr &parsedExpr) {
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (!evaluate::IsConstantExpr(*expr)) {  // C875,C881
+      context_.Say(parsedExpr.source, "Subscript must be a constant"_err_en_US);
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply lib/Semantics/check-data.cpp]: And perhaps a fold.
std::apply([&](const auto& ...x){(..., CheckIfConstantExpr(x));}, triplet.t);

File: test/Semantics/data03.f90 [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      DATA(b(x), i = 1, 5) / 5 * 1 /
+      !C881 
+      !OK: Correct use
+      DATA(nums % numbers(i), i = 1, 5) / 5 * 1 /
+      !C881
+      !OK: Correct use
+      DATA((d(i, j), i = 1, 10), j = 1, 10) / 100 * 1 /
+      !C881
+      !OK: Correct use
+      DATA(d(i, 1), i = 1, 10) / 100 * 1 /
```
Comment (kiranchandramohan) [inline test/Semantics/data03.f90]: Incorrect repeat?

**Comment by anchu-rajendran** [comment]
Thank you @tskeith, @kiranchandramohan for the comments. I have addressed your concerns. But build is failing only on arm64 gcc on a file which I have not edited. 

```
[36/190] Building CXX object lib/Parser/CMakeFiles/FortranParser.dir/debug-parser.cpp.o | 1733s
-- | --
5217 | [37/190] Building CXX object lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o | 1767s
5218 | FAILED: lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o | 1768s
5219 | /usr/local/bin/g++   -I../include -Iinclude -isystem ../llvm-project/install/include -UNDEBUG -std=c++17 -fno-rtti -fno-exceptions -pedantic -Wall -Wextra -Werror -Wcast-qual -Wimplicit-fallthrough -Wdelete-non-virtual-dtor -O2   -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MT lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o -MF lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o.d -o lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o -c ../lib/Evaluate/fold-real.cpp | 1768s
5220 | g++: fatal error: Killed signal terminated program cc1plus | 1768s
5221 | compilation terminated. | 1768s
5222 | [38/190] Building CXX object lib/Evaluate/CMakeFiles/FortranEvaluate.dir/variable.cpp.o
```

File: test/Semantics/data03.f90 [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      DATA(a[i], i = 1, 5) / 5 * 1 /
+      !C875
+      !ERROR: Data object variable must be a designator
+      DATA f(1) / 1 / 
+      !C875
+      !ERROR: Subscript must be a constant
+      DATA b(ind) / 1 /
+      !C875
+      !ERROR: Subscript must be a constant
+      DATA name( : ind) / 'Ancd' /
```
[the remark this answers is not in the document]
Comment (anchu-rajendran) [reply test/Semantics/data03.f90]: Done. Thank you.

File: lib/Semantics/check-data.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (anchu-rajendran) [reply lib/Semantics/check-data.cpp]: Done. Thank you.

File: test/Semantics/data03.f90 (hunk 1 above)
Comment (anchu-rajendran) [reply test/Semantics/data03.f90]: I mentioned it as Data object must not be a function reference.

File: test/Semantics/data03.f90 (hunk 3 above)
Comment (anchu-rajendran) [reply test/Semantics/data03.f90]: Done.

**Comment by anchu-rajendran** [comment]
> Thank you @tskeith, @kiranchandramohan for the comments. I have addressed your concerns. But build is failing only on arm64 gcc on a file which I have not edited.
> 
> ```
> [36/190] Building CXX object lib/Parser/CMakeFiles/FortranParser.dir/debug-parser.cpp.o | 1733s
> -- | --
> 5217 | [37/190] Building CXX object lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o | 1767s
> 5218 | FAILED: lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o | 1768s
> 5219 | /usr/local/bin/g++   -I../include -Iinclude -isystem ../llvm-project/install/include -UNDEBUG -std=c++17 -fno-rtti -fno-exceptions -pedantic -Wall -Wextra -Werror -Wcast-qual -Wimplicit-fallthrough -Wdelete-non-virtual-dtor -O2   -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MT lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o -MF lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o.d -o lib/Evaluate/CMakeFiles/FortranEvaluate.dir/fold-real.cpp.o -c ../lib/Evaluate/fold-real.cpp | 1768s
> 5220 | g++: fatal error: Killed signal terminated program cc1plus | 1768s
> 5221 | compilation terminated. | 1768s
> 5222 | [38/190] Building CXX object lib/Evaluate/CMakeFiles/FortranEvaluate.dir/variable.cpp.o
> ```

I pulled in recent changes and the build passed. I am still not sure why it failed.

File: lib/Semantics/check-data.cpp [hunk 5]
```diff
@@ -10,6 +10,59 @@
...
+template<typename T> void DataChecker::CheckIfConstantSubscript(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    if (!evaluate::IsConstantExpr(*expr)) {  // C875,C881
+      context_.Say(parser::FindSourceLocation(x),
+          "Data object must have constant bounds"_err_en_US);
+    }
+  }
+}
+
+void DataChecker::CheckSubscript(const parser::SectionSubscript &subscript) {
```
Comment (klausler) [inline lib/Semantics/check-data.cpp]: Would it be easier or cleaner to perform expression analysis on the variable and then apply these checks to the parsed expression?  That will catch more errors (e.g., rank problems) and we'll probably want to use type-checked expressions when converting the `DATA` statement into llvm.

Comment (anchu-rajendran) [reply lib/Semantics/check-data.cpp]: @klausler ,  I am not sure about this. I will try to implement this way. I believe you are asking me to call expression analysis on parser::Variable and to implement these checks when it is analysed. Please correct me if I am wrong. If it is so, DataImpliedDo objects are either Arrayrefs or Structure Components and they do not cover all variables as per parsing rule (They will not be parsed as parser::Variable). I can look into analysing dataref (which is called already for dataImpliedDo) but I dont think it helps in catching rank related issues.

Comment (klausler) [reply lib/Semantics/check-data.cpp]: Did you try this?  You requested another review, but I don't see any changes in the direction that I suggested.

Comment (anchu-rajendran) [reply lib/Semantics/check-data.cpp]: > Did you try this? You requested another review, but I don't see any changes in the direction that I suggested.

Sorry. My Mistake. I am trying to implement. I will soon notify with the change.

Comment (sscalpone) [reply lib/Semantics/check-data.cpp:22]: @anchu-rajendran The lowering of DATA statements to FIR would like the values in a data statement to have been evaluated.  I believe this work is in the scope of what you are doing.  Do you have an estimate for when you might push an update?

Comment (anchu-rajendran) [reply lib/Semantics/check-data.cpp:22]: > @anchu-rajendran The lowering of DATA statements to FIR would like the values in a data statement to have been evaluated. I believe this work is in the scope of what you are doing. Do you have an estimate for when you might push an update?

Hi Steve, I will push an update in a week.

Comment (anchu-rajendran) [reply lib/Semantics/check-data.cpp:22]: > @anchu-rajendran The lowering of DATA statements to FIR would like the values in a data statement to have been evaluated. I believe this work is in the scope of what you are doing. Do you have an estimate for when you might push an update?

The value evaluation of data statements is currently not under the scope of these semantic checks. Could you please share an example where data evaluation is necessary. I will look into the same and update.

**Comment by anchu-rajendran** [comment]
Closing the pull request for now and will reopen in phabricator.

**Pull request closed without merging**